docs: add get started page for server directory (#22154)
nuxt/nuxt: docs/1.getting-started/6.server-directory.md

---
navigation.icon: uil:laptop-connection
description: Using server directory to create middlewares, plugins and API easily.
---

# Server Directory

> Nuxt scans files under server directory to create middlewares, plugins and API routes to create features to use inside components or pages.

Reviewer: I think that we could add a first paragraph explaining what is the server directory and how it could be useful

Author: How is newly added paragraph? I tried to explain what is it for but might need help about it.

Reviewer: Will check it.

> Each file under these folders must return a function defined with `defineEventHandler()`.

Reviewer: ```suggestion
Each file under these folders must return a default export `defineEventHandler()`.

::aler{type="info"}
You can learn more about this function the [Nitro documentation](https://nitro.unjs.io/guide/routing).
::
```

## Server Routes

Anything under `~/server/api` folder will create API routes to fetch data within API route.

```ts [server/api/hello.ts]
export default defineEventHandler(() => {
    return {
      greetings: 'traveler'
    }
})
```

```vue [components/Greet.vue]
<script setup lang="ts">

> const {data} = await useFetch('/api/hello')

Reviewer: ```suggestion
const { data } = await useFetch('/api/hello')
```

> </script>

Reviewer: ```suggestion
</script>

<template>
  <div>
     {{ data }}
   </div>
</template>
```

I think that added a full example could help people to copy and paste to understand usage.

```

> This server route will return a data object that contains an object with content of `{greetings: 'traveler'}`.

Reviewer: ```suggestion
This server route will return a data object that contains an object with content of `{ greetings: 'traveler' }` printed in the template tag.
```

---
navigation.icon: uil:laptop-connection
description: Using server directory to create middlewares, plugins and API easily.
---

# Server Directory

> Server directory can be used for creating API routes to implement reusable data fetching from remote server easily, middlewares for intercepting requests and responses and plugins to hook into lifecycle hooks.

Reviewer: ```suggestion
Server directory can be used for creating API routes to implement data fetching from remote server, accessing database or performing actions on the server before returning some data.
```

Each file under these folders must return a default export `defineEventHandler()`.

## Server Routes

Anything under `~/server/api` folder will create API routes to fetch data within API route.

```ts [server/api/hello.ts]
export default defineEventHandler(() => {
    return {
      greetings: 'traveler'
    }
})
```

```vue [components/Greet.vue]
<script setup lang="ts">
const { data } = await useFetch('/api/hello')
</script>

<template>
  <div>
     {{ data }}
   </div>
</template>
```

This server route will return a data object that contains an object with content of `{greetings: 'traveler'}` printed in the template tag.

Server route file will automatically match with `GET` method and it can be changed by adding method names to file name such as: `.post`, `.put`, `.delete`.
Routes may accept parameters, body and query parameters to get some values.

:ReadMore{link="/docs/guide/directory-structure/server#matching-route-parameters"}

## Server Middleware

> Nuxt will read anything under `~/server/middleware` folder and create middlewares for you to tweak request information.

Reviewer 2: ```suggestion
Nuxt will read anything under `~/server/middleware` folder and create middleware for you to tweak request information.
```

```ts [server/middleware/log.ts]
export default defineEventHandler((event) => {
  event.context.auth = { user: 123 }
})
```

:ReadMore{link="/docs/guide/directory-structure/server#server-middleware"}

> ## Server Plugins

Reviewer: Maybe we could also add a section about the `utils` directory where content inside is auto-imported.

Reviewer 2: Good idea!

Author: Added basic explanation for `~/server/utils`. Haven't used it yet properly so it might be false/lacking.

Reviewer: Currently adding it to the Nitro documentation https://github.com/unjs/nitro/pull/1451. Will review your change.

---
navigation.icon: uil:laptop-connection

> description: Using server directory to create middlewares, plugins and API easily.

Reviewer 3: I feel like this should be less about the server _directory_ which is covered in https://nuxt.com/docs/guide/directory-structure/server and more about the _concepts_ of how Nuxt + Nitro inter-operate.

What do you think?

Server route file will automatically match with `GET` method and it can be changed by adding method names to file name such as: `.post`, `.put`, `.delete`.
Routes may accept parameters, body and query parameters to get some values.

:ReadMore{link="/docs/guide/directory-structure/server#matching-route-parameters"}

## Server Middleware

Nuxt will read anything under `~/server/middleware` folder and create middleware for you to tweak request information.

```ts [server/middleware/log.ts]
export default defineEventHandler((event) => {
  event.context.auth = { user: 123 }
})
```

:ReadMore{link="/docs/guide/directory-structure/server#server-middleware"}

## Server Plugins

Any files under `~/server/plugins` folder will be registered as Nitro Plugin for changing [Nitro](https://nitro.unjs.io/)'s behavior.

:ReadMore{link="/docs/guide/directory-structure/server#server-plugins"}

## Server Utils

[unjs/h3](https://github.com/unjs/h3) which powers server routes has many useful helpers that you can use inside server routes.

:ReadMore{link="https://www.jsdocs.io/package/h3#package-index-functions" title="Available H3 Request Helpers"}

> Incase of you want to have your own helper functions to solve your problem, they can be places under `~/server/utils` folder and they will be imported automatically for you to use.

Reviewer: ```suggestion
Incase of you want to have your own helper functions to solve your problem, they can be places under `~/server/utils` folder and they will be imported automatically for you to use like composables.
```
